cmd/admin-revoker/main.go

```diff
@@ -211,7 +240,129 @@ func (r *revoker) revokeByReg(ctx context.Context, regID int64, reasonCode revoc
 }
 
 func (r *revoker) revokeMalformedBySerial(ctx context.Context, serial string, reasonCode revocation.Reason) error {
-	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode)
+	return r.revokeCertificate(ctx, core.Certificate{Serial: serial}, reasonCode, false)
+}
+
+// blockByPrivateKey blocks future issuance for certificates with a public key
+// matching the SPKI hash of the provided private key. Reason code must be 1
+// (Key Compromise). Keys passed to this function will be validated as a pair
+// before any actions are taken. This method does not revoke any certificates
+// directly; however 'bad-key-revoker', which references the 'blockedKeys'
+// table, will eventually revoke certificates with a matching SPKI hash.
+func (r *revoker) blockByPrivateKey(ctx context.Context, privateKey crypto.Signer, reasonCode revocation.Reason) error {
+	if reasonCode < 0 || reasonCode != 1 {
+		panic(fmt.Sprintf("Invalid reason code: %d", reasonCode))
+	}
+
+	err := verifyPrivateKey(privateKey)
+	if err != nil {
+		return err
+	}
+
+	spkiHash, err := getPublicKeySPKIHash(privateKey.Public())
+	if err != nil {
+		return err
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return err
+	}
+
+	req := &sapb.AddBlockedKeyRequest{
+		KeyHash:   spkiHash,
+		Added:     r.clk.Now().UnixNano(),
+		Source:    "admin-revoker",
+		Comment:   fmt.Sprintf("blocked by %s", u),
+		RevokedBy: 0,
+	}
+
+	_, err = r.sac.AddBlockedKey(ctx, req)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// revokeByPrivateKey revokes all certificates with a public key matching the
+// SPKI hash of the provided private key. Reason code must be 1 (Key
+// Compromise). Keys passed to this function will be validated as a pair before
+// any actions are taken. The provided key will not be added to the
+// 'blockedKeys' table, this is done to avoid a race between 'admin-revoker'
+// and 'bad-key-revoker'.
```

Comment:
What do you think about having this function call `blockByPrivateKey` after it's done with all of the revocations, just so we don't have to rely on a human running the two commands back-to-back?

edit: I see that `main` calls these two functions back-to-back, so a human doesn't have to. I'd still consider doing it here, so that the blocking step absolutely definitely can't be skipped.

Reply:
I originally wrote it this way, however I removed it so that I could test that blocking wouldn't take place until all revocations had been completed successfully. Is there another way that I should write the test such that both could be true?

Reply:
Ooh good point. I don't have a good solution for that. You could introduce a channel that the revocation loop waits on and which the tests use to strategically stop and start the loop. But that'd be a lot of extra infrastructure and complication for this. In that case I'd probably just augment this comment with something to the effect of "You MUST call blockByPrivateKey after calling this function, on pain of violating the BRs."